### PR TITLE
Setup vendored libbpf and bcc

### PR DESCRIPTION
**Pull request by BurntBrunch**
This is a stepping stone to the work described in #2334. I pulled it forward in my schedule as it will make landing #2321 with clean CI possible.

Note that we can't just import libbpf - old bcc versions will not compile against new libbpf versions, so we need to pair them together. 

After a lot of iteration, I settled on a custom-known-prefix approach. Namely, USE_SYSTEM_BPF_BCC gates whether we add build-libs/ as a cmake prefix.

The reason this is superior to ExternalProject and other cmake-based approaches is that it allows us to keep the symbol existence checks untouched. (Symbol checks happen at generation time but ExternalProjects are built at build time, so the checks will always fail/run against system libraries.)

A new script ./build-libs.sh can build libbpf and bcc and install them into build-libs. This script is then setup to run from build.sh in CI.

The one remaining item to fix is that embedded builds on xenial do not have a linux/btf.h header at all, which breaks libbpf symbol existence checks.

### CODE REVIEW

**Review by viktormalik — COMMENTED** [review]
This looks good! Just one whitespace nit.

**Comment by viktormalik** [comment]
The new focal embedded build is failing to find the quay image since that one would only be published once bcb76973a73b3e67e252da65a24cd8923726fd18 is pushed to master. Could you perhaps move bcb76973a73b3e67e252da65a24cd8923726fd18 to a separate PR (which has already been tried in #2088, btw.) or does it require the previous commits as prerequisites?

**Comment by BurntBrunch** [comment]
I'll split out the xenial -> focal change, I missed that it checks for master before pushing.

**Review by viktormalik — APPROVED** [review]

**Comment by viktormalik** [comment]
You need to add `ENV DEBIAN_FRONTEND=noninteractive` to `docker/Dockerfile.ubuntu-glibc`, otherwise the embedded Focal build gets stuck.

**Comment by BurntBrunch** [comment]
CI error on the PR is just a transient apt error. [Embedded builds](https://github.com/BurntBrunch/bpftrace/actions/runs/3230113112) and [build_test](https://github.com/BurntBrunch/bpftrace/actions/runs/3230113107) passed in my repo.

**Comment by BurntBrunch** [comment]
CI error on the PR is just a transient apt error. [Embedded builds](https://github.com/BurntBrunch/bpftrace/actions/runs/3230113112) and [build_test](https://github.com/BurntBrunch/bpftrace/actions/runs/3230113107) passed in my repo.

**Review by ajor — COMMENTED** [review]
Is the ultimate goal to have users/developers building with these vendored libraries or are they exclusively for use in CI?

- The option USE_SYSTEM_BPF_BCC is currently a bit misleading because it defaults to OFF but then we still build using the system libraries (except in CI)
- If the build-libs.sh isn't intended to be run by users it can be hidden away somewhere other than the top-level directory, where users might be tempted to run it
- If build-libs.sh is intended to be run by users then it'd be nicer to have its outputs placed in the build directory, out of tree

It looks a bit odd to have a couple of empty git submodules in the top level source directory on my checkout. Can they be moved into a subdirectory out or sight, or as an alternative to Git submodules, the CMake FetchContent function seems like it'd be a good fit. Unlike ExternalProject it runs during configure time so would still work with check_symbol_exists. Then we wouldn't have these empty directories sitting around in our source tree at all.

**Comment by BurntBrunch** [comment]
> Is the ultimate goal to have users/developers building with these vendored libraries or are they exclusively for use in CI?

They're for developers and CI primarily.
 
> * The option USE_SYSTEM_BPF_BCC is currently a bit misleading because it defaults to OFF but then we still build using the system libraries (except in CI)

When USE_SYSTEM_BPF_BCC is OFF, it does prefer the locally built libraries if they're available. I can make it only use the local path, if it would make things clearer.

> * If the build-libs.sh isn't intended to be run by users it can be hidden away somewhere other than the top-level directory, where users might be tempted to run it

Define "users." If we're talking about people building bpftrace from source, then they do need to run it, as bpftrace starts depending on the versions of libbpf and bcc that it vendors (see #2370; the full plan is in #2334).

> * If build-libs.sh is intended to be run by users then it'd be nicer to have its outputs placed in the build directory, out of tree

Outputs are in build-libs. I'm not sure how to make cmake look at a non-source-dir-based path but I can try. 

> It looks a bit odd to have a couple of empty git submodules in the top level source directory on my checkout.

They shouldn't be empty, you haven't updated them. `git submodule update --recursive`.

> the CMake FetchContent function seems like it'd be a good fit

It's actually not a good fit - FetchContent would at most work for bcc, which also uses cmake, but it would pull the bcc build into the bpftrace build. The final output of FetchContent is meant to be a cmake-using subdirectory, not a fully built library. It doesn't work for libbpf and I'm reasonably sure it won't work with the symbol checks either.

**Review by ajor — CHANGES_REQUESTED** [review]
> Define "users." If we're talking about people building bpftrace from source, then they do need to run it, as bpftrace starts depending on the versions of libbpf and bcc that it vendors (see #2370; the full plan is in #2334).

I mean anyone who's building bpftrace from source. I say users because not everyone who builds bpftrace from source is a developer - they might just not be on a distro that has a package and don't want to do any coding themselves.

The other ".sh" files in the project root are intended to be run manually by users to run certain builds - see https://github.com/iovisor/bpftrace/blob/master/INSTALL.md#using-docker

I'm not sure I'm understanding completely, but if people need to run this to build bpftrace then it should be documented in INSTALL.md

> Outputs are in build-libs. I'm not sure how to make cmake look at a non-source-dir-based path but I can try.

The CMake variable `CMAKE_BINARY_DIR` will point to the user's chosen build directory, so that'd be a good place to build these libs too.


> They shouldn't be empty, you haven't updated them. `git submodule update --recursive`.

That's right, but I didn't realise that I needed to check them out. Is this required? It should be documented in INSTALL.md if so.

File: src/ast/CMakeLists.txt
Comment (BurntBrunch) [reply src/ast/CMakeLists.txt:51]: Nothing in particular, they're just now a requirement on the newer alpine version for the static build to work. There are references from clang to these libs that would be left unresolved otherwise.

I really want to find a better way to express the static llvm build, all these details leaking to us is not great.

File: .gitmodules
Comment (BurntBrunch) [reply .gitmodules:1]: There are things that are APIs that disappear in the 1.0 release, which in turn break the chosen version of bcc. Changing this is certainly possible but let's do it one step at a time.

**Comment by BurntBrunch** [comment]
Changes since last round: 
 * Documented build-libs.sh usage in INSTALL.md.
 * Made the cmake build _require_ the vendored libraries when USE_SYSTEM_BPF_BCC=0. This also errors out with a message telling you to run build-libs.sh. Overall this is actually nicer as you can swap between system and vendored libraries even if you have already built the vendored ones.
 * Moved the build-libs.sh output to $PWD/build-libs so parallel docker builds don't clash with each other
 * Cleaned up libbpf_version, bcc_ref

**Review by ajor — COMMENTED** [review]
Thanks, looks good to me!

The LGTM check is failing for this branch but I know nothing about it. Maybe something similar to [7d512019737374b6bb126a1f2ae805d54036257e](https://github.com/iovisor/bpftrace/pull/2265/commits/7d512019737374b6bb126a1f2ae805d54036257e) is needed? @viktormalik

**Comment by BurntBrunch** [comment]
Yeah, looks like we either need to have lgtm create the build folder and run build-libs.sh explicitly or somehow pass USE_SYSTEM_BPF_BCC=1 to its cmake invocation. I didn't see an obvious way to do that, so fully manual it is I think but I'll wait for @viktormalik's opinion.

Given LGTM's EOL, it might be a better idea to remove it and just migrate to the Github equivalent.

**Comment by viktormalik** [comment]
> Yeah, looks like we either need to have lgtm create the build folder and run build-libs.sh explicitly or somehow pass USE_SYSTEM_BPF_BCC=1 to its cmake invocation. I didn't see an obvious way to do that, so fully manual it is I think but I'll wait for @viktormalik's opinion.

There seems to be a way to tell LGTM how to execute CMake by [customizing the configure step](https://lgtm.com/help/lgtm/cpp-extraction#customizing-configure). But that would rely on the system libbpf which may be fairly old, so I'd go with the manual variant (basically something similar to what 7d512019737374b6bb126a1f2ae805d54036257e does).

**Comment by BurntBrunch** [comment]
I think I managed to make LGTM work. Took me a lot of trial and error to make it use the correct llvm package.

Unfortunately, it looks like the llvm-15 job is prone to failure (hasn't failed yet on the last push at the time of writing) with this error during the bcc build, which is bizarre to say the least.

```
[100%] Linking CXX shared library libbcc.so
/usr/bin/ld: /usr/lib/llvm-15/lib/libclangFrontend.a: member /usr/lib/llvm-15/lib/libclangFrontend.a(ASTConsumers.cpp.o) in archive is not an object
collect2: error: ld returned 1 exit status
```

**Comment by viktormalik** [comment]
> Unfortunately, it looks like the llvm-15 job is prone to failure (hasn't failed yet on the last push at the time of writing) with this error during the bcc build, which is bizarre to say the least.

**Comment by viktormalik** [comment]
Sorry for the unintended PR close (the "Close PR" button looks too much like a "Cancel comment" button).

> Unfortunately, it looks like the llvm-15 job is prone to failure (hasn't failed yet on the last push at the time of writing) with this error during the bcc build, which is bizarre to say the least.

I've also seen another strange CMake error (not this one, though) with LLVM 15. It disappeared after some time, so I guess that the apt package was fixed. This could be something similar.

**Review by ajor — APPROVED** [review]

**Comment by BurntBrunch** [comment]
I can't rerun the PR jobs but here is the LLVM-15 build succeeding on my fork - https://github.com/BurntBrunch/bpftrace/actions/runs/3324984829/jobs/5510027311

This does indeed point to faulty packages (or more likely CDNs).

The overall test failure is from a timeout - I think we should maybe make timeouts retry, as relying on them in a cloud environment (especially wall clock) is always a losing battle.

Anywho, I think this can be merged now, especially if someone can kick the PR builds.

**Pull request merged**
